hadoop-hdds/docs/content/design/decommissioning.md

```diff
@@ -0,0 +1,720 @@
+
+To support idempontent placement strategies we include the in-fligt replications in the `currentCount`: If there are one in-flight replication process and two replicas we won't start a new replication command unless the original command is timed out.
+
+The timeout is configured with `hdds.scm.replication.event.timeout` and the default value is 10 minutes.
+
+More preciously the current algorithm is the following:
+
+```
+Replica count = expectedCount - healthy - inFlight
+```
+
+### The proposed solution
+
+To support the notion that a user can provide hints to the replication model, we propose to add two variables to the current model.
+
+In the new model, we propose to break the `currentCount` into the three separate groups. That is _Healthy nodes_, _Maintenance nodes_, and _Decommission nodes_. The new model replaces the currentCount with these three separate counts. The following function captures the code that drives the logic of computing Replica counts in the new model. The table below discusses the input and output of this model very extensively.
+
+```java
+/**
+ * Calculate the number of the missing replicas.
+ * 
+ * @return the number of the missing replicas. If it's less than zero, the container is over replicated.
+ */
+int getReplicationCount(int expectedCount, int healthy, 
+   int maintenance, int inFlight) {
+
+   //for over replication, count only with the healthy replicas
+   if (expectedCount < healthy) {
+      return expectedCount - healthy;
+   }
+   
+   replicaCount = expectedCount - (healthy + maintenance + inFlight);
+
+   if (replicaCount == 0 && healthy < 1) {
+      replicaCount ++;
+   }
+   
+   //over replication is already handled
+   return Math.max(0, replicaCount);
+}
+
+```
+
+We also need to specify two end condition when the DECOMMISSIONING node can be moved to the DECOMMISSIONED state or the ENTERING_MAINTENANCE mode can be moved to the IN_MAINTENANCE state.
+
+The following conditions should be true for all the containers and all the containers on the specific node should be closed.
+
+From DECOMMISSIONING to DECOMMISSIONED:
+
+ * There are at least one healthy replica
+ * There are at most one missing replica
+
+Which means that node can be decommissioned if: 
+ 
+ * all the containers with replication factor THREE have at least *one replica* on a HEALTHY nodes (minimum.live.replicas)
+ * all the containers with replication factor THREE have at least *three* replicas on HEALTHY/ENTERING_MAINTENENCE/IN_MAINTENANCE nodes (minimum.replicas)
+ * all the containers with replication factor ONE have on replica on a HEALTHY node.
+
+
+From ENTERING_MAINTENANCE to IN_MAINTENANCE:
+
+ * There are at least one healthy replicas
```

Comment:
I mentioned this earlier, but we should make this a configurable number, likely 1 or 2.

Reply:
Sure, it's added to here...